lib/src/c.rs

```diff
@@ -85,301 +181,474 @@ fn didkit_generate_secp384r1_key() -> Result<*const c_char, Error> {
+/// data proof. Input should be a C string containing the key as a JWK. The JWK should contain
+/// public key material; private key parameters are ignored. On success, this function returns a
+/// newly-allocated C string containing the `verificationMethod` URI. On failure, `NULL` is
+/// returned; the error message can be retrieved using [`didkit_error_message`].
+#[no_mangle]
+pub extern "C" fn didkit_key_to_verification_method(
     method_pattern_ptr: *const c_char,
     key_json_ptr: *const c_char,
-) -> Result<*const c_char, Error> {
-    let method_pattern = unsafe { CStr::from_ptr(method_pattern_ptr) }.to_str()?;
-    let key_json = unsafe { CStr::from_ptr(key_json_ptr) }.to_str()?;
-    let key: JWK = serde_json::from_str(key_json)?;
+) -> *const c_char {
+    string_from_raw_ptr(method_pattern_ptr)
+        .and_then(
+            |method_pattern|
+            from_json_raw_ptr::<JWK>(key_json_ptr)
+                .map(|jwk| (method_pattern, jwk))
+        )
+        .and_then(|(method_pattern, jwk)| key_to_verification_method(&method_pattern, jwk))
+        .and_then(to_char_raw_ptr)
+        .unwrap_or_else(stash_err)
+}
+
+fn key_to_verification_method(
+    method_pattern: &str,
+    jwk: JWK
+) -> Result<String, Error> {
     let did_method = DID_METHODS
         .get(method_pattern)
         .ok_or(Error::UnknownDIDMethod)?;
     let did = did_method
-        .generate(&Source::Key(&key))
+        .generate(&Source::Key(&jwk))
         .ok_or(Error::UnableToGenerateDID)?;
     let did_resolver = did_method.to_resolver();
     let rt = runtime::get()?;
-    let vm = rt
-        .block_on(get_verification_method(&did, did_resolver))
-        .ok_or(Error::UnableToGetVerificationMethod)?;
-    Ok(CString::new(vm)?.into_raw())
+    rt.block_on(get_verification_method(&did, did_resolver))
+        .ok_or(Error::UnableToGetVerificationMethod)
 }
 
+
 /// Issue a Verifiable Credential. Input parameters are JSON C strings for the unsigned credential
 /// to be issued, the linked data proof options, and the JWK for signing.  On success, the
 /// newly-issued verifiable credential is returned as a newly-allocated C string.  The returned
-/// string should be freed using [`didkit_free_string`]. On failure, `NULL` is returned, and the error
-/// message can be retrieved using [`didkit_error_message`].
-#[didkit_macros::c_export(wrap = "ccchar_or_error")]
-fn didkit_vc_issue_credential(
+/// string should be freed using [`didkit_free_string`]. On failure, `NULL` is returned, and the
+/// error message can be retrieved using [`didkit_error_message`].
+#[no_mangle]
+pub extern "C" fn didkit_vc_issue_credential(
     credential_json_ptr: *const c_char,
     proof_options_json_ptr: *const c_char,
     key_json_ptr: *const c_char,
     context_loader_ptr: *const c_char,
-) -> Result<*const c_char, Error> {
+) -> *const c_char {
+    from_json_raw_ptr::<VerifiableCredential>(credential_json_ptr)
+        .and_then(
```

Comment:
I find these `and_then` hard to read, maybe it would be better to keep this code in `vc_issue_credential` to be able to use variables and `?`

Reply:
Sure thing.

The macro package is needed by the java code as well, so I can't remove it yet.   I'll get the clippy errors this afternoon.

Reply:
@sbihel Please check didkit_key_to_did and didkit_vc_issue_credential and see if you like that approach better.

Reply:
> The macro package is needed by the java code as well, so I can't remove it yet.

I meant the `c_export` macro

> I think the functions like vc_issue_credential belong in a different module

Yes absolutely

>  I've addressed them in the file I touched

Yes that's what I meant

> Please check didkit_key_to_did and didkit_vc_issue_credential and see if you like that approach better.

I think it's a bit convoluted, can we just have the deserialisation with the business logic for now like it was?